### ISSUE TITLE
Unable to reproduce Web30K results of GSF (Groupwise Ranking)

### ISSUE THREAD

**Issue by fangkuann** [tensorflow/ranking#148]
Hi， 
   
   We are working on a research paper which is related to [Groupwise Ranking](https://arxiv.org/pdf/1811.04415v3.pdf). We intend to use [Groupwise Ranking] as one of our main baseline methods.  However, we found it is hard to reproduce GSF results on the MSLR Web30K dataset.  
    
We can not reproduce any of the GSF alternatives' results reported on the paper. To make clear if we are going anything wrong, we list the reproducing parameters of GSF(1), which is a GroupWise ranking function with group size = 1.
    
We use the example code[ tf_ranking_libsvm.py](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_libsvm.py) to train the GSF(1) model. According to the original paper, we use the same hyperparameters reported on it, 

```
data=Web30K-Fold1 (With all-zero labels' query removed)
learning_rate=0.005
train_batch_size=128
num_train_steps=30000
act_function=ReLu
each_layer=(FullConnect-BatchNorm-Relu)
input_process=BatchNorm
hidden_layer_dims=[64, 32, 16]
num_features=136
list_size=200
group_size=1
loss=softmax_loss
dropout_rate=0.0
```

After 30000 steps' training, the NDCG@5 on test set is 39.56517, which is much lower than the result on paper: 40.4. 

Is there any other parameters we ignored, we are not certain about some hyperparameters not listed on the paper:

-  The paper describes the reported data are run 10 times and average over these results. What is the result used in each time's running for final average, the best Model during the 30000 step's training or the last model of 30000 steps' training?

- Did GroupWise ranking use dropout technique, if yes, what is the way it applied

- Did learning rate decay/ warmup be applied during training? 

- Did feature normalization using global feature mean/variance(not BN) be applied before fed into the score function?
- What're the BN parameters used in GSF?
- Since the TF-Ranking uses padding technique to make the dynamic query doc size available, what's the list_size used in producing GroupWise-ranking, and will padded record influence BatchNormalization's computation? 

We will be sincerely grateful if any support is given. 

Thanks

**Comment by sbruch** [comment]
Thanks for reaching out to us! I'll try to answer your questions based on what I understand from the paper and speaking with a few of the authors.

Q1. The best model based on performance on the validation set.
Q2. I don't believe it used dropout for Web30K experiments.
Q3. No learning rate decay/warmup was applied.
Q4. Only batch normalization was used between each layer as well as over the input layer.
Q5. Default values.
Q6. During training 'list_size' (as you correctly pointed out) is set to 200, but for evaluation it is set to the maximum list size (e.g., 1300).

One thing you didn't ask but may be relevant is that the paper uses Adagrad to optimize the ranking loss.

**Comment by sbruch** [comment]
As a side note, it may be better to direct further questions about the paper to the corresponding author listed on the paper. Issues on Github are generally questions about the library, not any particular publication that uses the library.

**Comment by eggie5** [comment]
Why didn't I think to reset the `LIST_SIZE` during testing!?

**Comment by fangkuann** [comment]
> Thanks for reaching out to us! I'll try to answer your questions based on what I understand from the paper and speaking with a few of the authors.
> 
> Q1. The best model based on performance on the validation set.
> Q2. I don't believe it used dropout for Web30K experiments.
> Q3. No learning rate decay/warmup was applied.
> Q4. Only batch normalization was used between each layer as well as over the input layer.
> Q5. Default values.
> Q6. During training 'list_size' (as you correctly pointed out) is set to 200, but for evaluation it is set to the maximum list size (e.g., 1300).
> 
> One thing you didn't ask but may be relevant is that the paper uses Adagrad to optimize the ranking loss.

Thanks for the detailed response! 

We missed an important point -- add query weight by the sum of the relevance grade of its documents. 

![image](https://user-images.githubusercontent.com/5515861/70612018-19d40a00-1c41-11ea-8d3b-37e5815c34df.png)

We implement it by using **_weights_feature_name_** param in **_tfr.losses.make_loss_fn_** and create a feature named 'query_weight' filled with all same value of the sum of documents' grade score in a query.

After this process, in PairGSF(PairwiseLogisticLoss + group_size=1), the ndcg@5 metric caught up a little, from 39.5 -> 39.8, but still a little distance to the paper's result(40.4).  Then we set learning_rate from 0.005 to 0.015, the final result came up to the reported version. 

However, then we convey these parameters(query weight + lr=0.015/0.005) to the SoftMax-Loss, the performance even degraded.  And we remove the query weight technique and enlarge the learning rate to 0.2, the result could catch up. 

Finally, in our experiment, there are mainly two conclusions:

1. add query weight(if we did it properly), makes the model perform better in PairwiseLogisticLoss, but worse in SoftmaxLoss
2. enlarge the learning rate could make results coincident to the of the paper.

Does it make sense? Or anything still went wrong in our experiment.  

I also sent an email to the corresponding author and leave the same topic there.

**Comment by sbruch** [comment]
Let's continue this conversation over email, where the first author can comment as well. I'll close this issue for the time being. But to summarize, if you're interested in exactly reproducing the results with the same set of hyperparameters reported in the paper, you'll likely also need to use the same Tensorflow and TFRanking revisions. Having said that, you should be able to reproduce results reasonably close to the reported metrics by tuning the hyperparameters yourself -- which is what you appear to have done.

**Comment by stanpcf** [comment]
Hi, 
I'm confused at how to combine lambdaMart and GSF in this paper.
![image](https://user-images.githubusercontent.com/13925796/112842642-5d6fb300-90d4-11eb-8738-0ec44adab575.png)

In my experiment, the scores which is output by nn and gbdt is not the same scale, add this two score not a good idea.

We will be sincerely grateful if any support is given.
Thanks
